radixdlt-java/src/main/java/com/radixdlt/client/application/RadixApplicationAPI.java

```diff
@@ -841,12 +841,32 @@ public Result execute(Action action) {
 	 * NOR atomic (i.e. if an action fails, all previous actions to that would still have occurred).
 	 *
 	 * @param actions the action to execute sequentially
-	 * @return completion status of all of the actions
+	 * @return list of results
 	 */
-	public Completable executeSequentially(Action... actions) {
-		Completable completable = Observable.fromIterable(Arrays.asList(actions))
-			.concatMapCompletable(a -> buildDisconnectedResult(a).connect().toCompletable()).cache();
-		completable.subscribe();
-		return completable;
+	public List<Result> executeSequentially(Action... actions) {
+		return executeSequentially(Arrays.asList(actions));
+	}
+
+	/**
+	 * Executes actions sequentially. If an action fails, then the completable this method
+	 * returns will call onError immediately. Note that this method is NEITHER idempotent
+	 * NOR atomic (i.e. if an action fails, all previous actions to that would still have occurred).
+	 *
+	 * @param actions the action to execute sequentially
+	 * @return list of results
+	 */
+	public List<Result> executeSequentially(List<Action> actions) {
+		List<Result> results = actions.stream().map(this::buildDisconnectedResult).collect(Collectors.toList());
+
+		Observable.fromIterable(results)
+			.concatMap(result -> result.connect().toObservable())
+			.takeUntil(a ->
+				a instanceof SubmitAtomResultAction
```

Comment:
Hmm, it appears that this changes the behaviour of executeSequentially, but apparently this method isn't used anywhere, so it's probably fine.

Reply:
It's actually close to the same